webmin/os-eol-lib.pl

```diff
@@ -0,0 +1,253 @@
+                }
+        }
+else {
+        $eol_data->{'_eol'} =
+                &make_date($eol_data->{'_eol_timestamp'}, 1);
+        if ($eol_data->{'extendedSupport'}) {
+                $eol_data->{'_ext_eol'} =
+                        &make_date($eol_data->{'_ext_eol_timestamp'}, 1);
+                }
+        }
+
+# Is expired?
+my $expired = $eol_data->{'_eol_timestamp'} < time();
+$eol_data->{'_expired'} = $text{'os_eol_reached'} if ($expired);
+
+# Is expiring (in 3 months by default, unless configured otherwise)
+my $os_eol_before = $gconfig{'os_eol_before'} // 3;
+if (!$expired && $os_eol_before) {
+        my $expiring = $eol_data->{'_eol_timestamp'} < time() +
+                                60*60*24*30*$os_eol_before ? 1 : 0;
+        # Set the expiring message
+        if ($expiring) {
+                $eol_data->{'_expiring'} =
+                        $eol_data->{'_eol_in'} ?
+                                $text{'os_eol_reaching'} . " " .
+                                        $eol_data->{'_eol_in'} :
+                                $text{'os_eol_reaching2'};
+                }
+        # Set extended support expiring message (if available)
+        if ($eol_data->{'extendedSupport'}) {
+                my $expiring = $eol_data->{'_ext_eol_timestamp'} < time() +
+                                60*60*24*30*$os_eol_before ? 1 : 0;
+                if ($expiring) {
+                        $eol_data->{'_ext_expiring'} =
+                                $eol_data->{'_ext_eol_in'} ?
+                                        $text{'os_eol_reaching'} . " " .
+                                                $eol_data->{'_ext_eol_in'} :
+                                        $text{'os_eol_reaching2'};
+                        }
+                }
+        }
+return $eol_data;
+}
+
+# eol_update_cache()
+# Caches the use of EOL data for next 30 days
+# or until the next Webmin update
+sub eol_update_cache
+{
+my (%miniserv, $updated);
+# Is there valid cached data (cached for 30 days)
+if ($gconfig{'os_eol_expired'} || $gconfig{'os_eol_expiring'}) {
+        if ($gconfig{'os_eol_last'} &&
+            $gconfig{'os_eol_last'} < time() - 60*60*24*30) {
+                # Invalidate the cache
+                foreach my $key
+                        ('os_eol_none', 'os_eol_expired',
+                         'os_eol_expiring', 'os_eol_last',
+                         'os_eol_about', 'os_eol', 'os_eol_in',
+                         'os_ext_eol', 'os_ext_eol_in') {
+                        delete($gconfig{$key});
+                        }
+                $updated++;
+                }
+        else {
+                # Nothing to update, Webmin config has it all
+                return;
+                }
+        }
+&get_miniserv_config(\%miniserv);
+if (!$gconfig{'os_eol_none'} ||
+    ($miniserv{'server'} && $gconfig{'os_eol_none'} ne $miniserv{'server'})) {
```

Comment:
What is this checking the miniserv config for?

Reply:
It checks if `os_eol.json` was updated/changed after Webmin upgrade. It's useful in case we add or remove some EOL data for an OS.